### ISSUE TITLE
[API Proposal]: TimeSpan - Parse ISO8601 Durations

### ISSUE THREAD

**Issue by Foxtrek64** [dotnet/runtime#72064]
### Background and motivation

Currently I am integrating with a JSON REST API which returns its timespans formatted as [ISO8601 Durations](https://en.wikipedia.org/wiki/ISO_8601#Durations). Attempting to parse these values with `TimeSpan.Parse()` is currently not available. Parsing with `TimeSpan.ParseExact()` may work with a custom formatting string, but I have not had any luck and currently, no support for parsing groups larger than days exists.

### API Proposal

This proposal does not add any new APIs. It only modifies existing APIs.

1. Add constant "o" format specifier for `ParseExact()` and `TryParseExact()`. This is in reference to the same value used by DateTime and DateTimeOffset to serialize and deserialize ISO8601 timestamp formats.
2. Add support for parsing years and months using `y`, `yy`, `yyyy`, `yyyyy`, and `M`, `MM` respectively.
3. Optionally add support for `Parse()` and `TryParse()` if this proves not to be too complicated.

### API Usage

```csharp
string duration = "P3Y6M4DT12H30M5S";
var timespan = TimeSpan.ParseExact(duration, "o", null); // 3 years, 6 months, 4 days, 12 hours, 30 minutes, 5 seconds.
```

### Alternative Designs

Currently a viable workaround is to use `XmlConvert.ToTimeSpan(string duration)` and `XmlConvert.ToString(TimeSpan duration)`, however this lives in the `System.Xml` namespace and usually makes little sense to add, especially when working with JSON.

**Comment by Clockwork-Muse** [comment]
Being able to parse into a `TimeSpan` isn't going to solve any problems, it's immediately going to give you a new one: For any field larger than "Day", durations are only relative.  "Month", for instance, may be equivalent to 28, 29, 30, or 31 days.  You cannot reliably translate this to a `TimeSpan` and then add it to a `DateTime`/`DateTimeOffset`, you're going to be dealing with incorrect values.

We'd need to create a type that's the equivalent of the [`Period`](https://nodatime.org/3.1.x/api/NodaTime.Period.html) class from NodaTime/JSR310 to be able to actually handle this.
<hr />

As a side note, if you're actually dealing with date/time problems like this, you're probably better advised to just use NodaTime anyways, as there are a number of shortcomings in the BCL that it addresses (although not as many as there used to be).

<hr />

 Actually, technically "Day" is sometimes relative too, about once every 18 months, due to Leap Seconds.  Although most APIs I'm aware of ignore them anyways.

**Comment by tannergooding** [comment]
CC. @tarekgh, is this something we should look at providing?

**Comment by tarekgh** [comment]
I agree with @Clockwork-Muse, cannot parse duration format to a TimeSpan. We may consider supporting `duration` by introducing a new type. We just want to see a demand for that, and we can consider it.

**Comment by Foxtrek64** [comment]
> I agree with @Clockwork-Muse, cannot parse duration format to a TimeSpan. We may consider supporting `duration` by introducing a new type. We just want to see a demand for that, and we can consider it.

I'll go ahead and update the proposal to recommend the creation of a new type. This will help us better gauge demand.

**Comment by tthiery** [comment]
Please remove `System.Xml.XmlConvert.ToTimeSpan` as a *viable* design. It does only support normalized to days which is highly inaccurate when working with months or years. The parsed timespan is generally not accepted as a suitable (see #28862).

**Comment by WhatzGames** [comment]
Should it be considered to add some new new operators for working with TimeOnly and DateOnly?

```csharp
public static TimeOnly operator +(TimeOnly left, Duration right);
public static TimeOnly operator -(TimeOnly left, Duration right);

public static DateOnly operator +(DateOnly left, Duration right);
public static DateOnly operator -(DateOnly left, Duration right);

public static Duration operator -(DateOnly left, DateOnly right);
public static Duration operator -(TimeOnly left, TimeOnly right);
```

These were the ones that would make sense to me.
Calculating differences between Times or Dates.
Adding a Duration to a `DateOnly` would only add everything from Days, Weeks, Months and Years.
The `TimeOnly` then only adds the time components and would overflow back to zero

**Comment by mstgms** [comment]
Is there any progress? We also need can parse duration values which sents from clients in jsons like "PT7H0.301S" on our Financial APIs. 
+@muazcoban

**Comment by tarekgh** [comment]
We're currently occupied with higher-priority tasks. We'll address this at a later time. Thank you!

**Comment by Clockwork-Muse** [comment]
@mstgms - especially if you're dealing with many of the scenarios where date-based durations make sense, you may find the overall API from [NodaTime](https://nodatime.org/) more ergonomic for many tasks, which also includes a [`Period` type](https://nodatime.org/3.1.x/api/NodaTime.Period.html).

**Comment by julealgon** [comment]
When/if this is implemented, will the framework start adding overloads to timeout-related methods that take a `Duration` instead of a `TimeSpan`? Any other areas in the framework where having this type would be better than a `TimeSpan` currently?

Would this also potentially deprecate `TimeSpan` over time? Is there any specific reason someone would _not_ use a `Duration` instead of a `TimeSpan` in their domains?

**Comment by tarekgh** [comment]
> When/if this is implemented, will the framework start adding overloads to timeout-related methods that take a Duration instead of a TimeSpan? Any other areas in the framework where having this type would be better than a TimeSpan currently?

This depends on if it makes sense to have any API accept duration. That need to look at case by case to decide that.

> Would this also potentially deprecate TimeSpan over time? Is there any specific reason someone would not use a Duration instead of a TimeSpan in their domains?

I don't think we need to deprecate TimeSpan at all. TimeSpan and Duration are needed for different purposes. You may look at the blog https://thecontentauthority.com/blog/span-vs-duration describing the difference between the time span and the duration.
